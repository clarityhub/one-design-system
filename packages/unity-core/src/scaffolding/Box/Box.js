import styled from '@emotion/styled';
import { bool, oneOf } from 'prop-types';

const Box = styled.div`
	padding-top: ${({ withGutterTop, withGutter }) => withGutterTop || withGutter ? '1rem' : '0rem'};
	padding-bottom: ${({ withGutterBottom, withGutter }) => withGutterBottom || withGutter ? '1rem' : '0rem'};
	padding-left: ${({ withGutterLeft, withGutter }) => withGutterLeft || withGutter ? '1rem' : '0rem'};
	padding-right: ${({ withGutterRight, withGutter }) => withGutterRight || withGutter ? '1rem' : '0rem'};
	float: ${({ pull }) => pull ? pull : 'inherit'};
`;

Box.propTypes = {
	pull: oneOf(['left', 'right']),
	withGutter: bool,
	withGutterBottom: bool,
	withGutterLeft: bool,
	withGutterRight: bool,
	withGutterTop: bool,
};

Box.defaultProps = {
	withGutter: false,
	withGutterTop: false,
	withGutterBottom: false,
	withGutterLeft: false,
	withGutterRight: false,
};

export default Box;

