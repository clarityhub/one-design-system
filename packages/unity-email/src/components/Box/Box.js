import React, { Fragment } from 'react';
import {
	MjmlSpacer,
} from 'mjml-react';
import { node, bool, number, oneOfType } from 'prop-types';

const Box = ({
	children,
	marginTop = false,
	marginBottom = false,
}) => (
	<Fragment>
		{marginTop && <MjmlSpacer height={`${marginTop * 16}px`} />}
		{children}
		{marginBottom && <MjmlSpacer height={`${marginBottom * 16}px`} />}
	</Fragment>
);

Box.propTypes = {
	children: node,
	marginBottom: oneOfType([bool, number]),
	marginTop: oneOfType([bool, number]),
};

export default Box;

