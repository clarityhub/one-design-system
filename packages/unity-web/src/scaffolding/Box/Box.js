import React from 'react';
import styled from '@emotion/styled';
import { css } from '@emotion/core';
import { any, oneOf, oneOfType, shape, string, func, node } from 'prop-types';

import colors, { types } from '@clarityhub/unity-core/lib/colors';
import spacing from '@clarityhub/unity-core/lib/spacing';

const remMapper = (size) => {
	return `${spacing[size]}rem`;
};

const rectMapper = (top, right, bottom, left) => {
	return `${top ? remMapper(top) : 0} ${right ? remMapper(right) : 0} ${bottom ? remMapper(bottom) : 0} ${left ? remMapper(left) : 0}`;
};

const StyledBox = styled.div(
	css`
		display: flex;
	`,
	({ align }) => align && css`
		text-align: ${align};
	`,
	({ gap, direction }) => {
		if (gap) {
			if (direction === 'column') {
				return css`
					& > *:not(:first-of-type) {
						margin-top: ${remMapper(gap)};
					}
				`;
			} else {
				return css`
					& > *:not(:first-of-type) {
						margin-left: ${remMapper(gap)};
					}
				`;
			}
		}
	},
	({ flex }) => flex && css`
		flex: ${flex || 1};
	`,
	({ direction }) => css`
		flex-direction: ${direction || 'column'};
	`,
	({ pull }) => pull && css`
		float: ${pull ? pull : 'inherit'};
	`,
	({ padding }) => {
		const def = typeof padding === 'string' ? padding : 0;
		return padding && css`
			padding: ${rectMapper(padding.top || def, padding.right || def, padding.bottom || def, padding.left || def)};
		`;
	},
	({ margin }) => {
		const def = typeof margin === 'string' ? margin : 0;
		return margin && css`
			margin: ${rectMapper(margin.top || def, margin.right || def, margin.bottom || def, margin.left || def)};
		`;
	},
	({ background }) => background && colors[background] && css`
		background: ${colors[background].default};
	`,
	({ color }) => color && colors[color] && css`
		color: ${colors[color].default};
	`,
	({ type }) => type && types[type] && css`
		background: ${types[type].background.default};
		color: ${types[type].color.default};
	`,
);

const Box = ({
	as,
	a11yTitle,
	direction,
	gap,
	flex,
	margin,
	padding,
	pull,
	...props
}) => {
	return (
		<StyledBox
			as={as}
			aria-label={a11yTitle}
			direction={direction}
			flex={flex}
			gap={gap}
			pull={pull}
			margin={margin}
			padding={padding}
			{...props}
		/>
	);
};

// TODO refactor into shared proptypes
const sizesProp = oneOf([
	'xsmall',
	'small',
	'large',
	'xlarge',
	'medium',
]);

// TODO refactor into shared proptypes
const typesProp = oneOf([
	'primary',
	'danger',
	'success',
	'brand',
]);

const genericSizesProp = oneOfType([
	sizesProp,
	shape({
		top: sizesProp,
		bottom: sizesProp,
		left: sizesProp,
		right: sizesProp,
	}),
]);

/**
 * Hack for docz and react-docgen
 */
const BoxProps = () => <div />;
BoxProps.propTypes = {
	a11yTitle: string,
	align: oneOf(['left', 'right', 'center']),
	as: oneOfType([string, func, node]),
	background: string,
	color: string,
	direction: oneOf(['column', 'row']),
	flex: any,
	gap: sizesProp,
	margin: genericSizesProp,
	padding: genericSizesProp,
	pull: oneOf(['left', 'right']),
	type: typesProp,
};

Box.propTypes = BoxProps.propTypes;
Box.defaultProps = BoxProps.defaultProps;
export { BoxProps };

export default Box;

