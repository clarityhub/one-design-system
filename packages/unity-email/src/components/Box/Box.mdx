---
menu: Components
name: Box
---

import { Playground, PropsTable } from 'docz';
import Box from './Box';

# Email Box

```js
import Box from '@clarityhub/unity-email/components/Box';
```

## Prop Types

<PropsTable of={Box} />
