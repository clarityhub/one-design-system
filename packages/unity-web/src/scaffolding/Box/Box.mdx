---
name: Box
menu: Scaffolding
---

import { Playground, PropsTable } from 'docz';
import Box, { BoxProps } from './Box';
import Typography from '../../components/Typography';

# Box

Box is a primitive that you can use to space out elements on a page.

## \<Box /\> component

```js
import Box from '@clarityhub/unity-web/lib/scaffolding/Box';
```

## Examples

### Colored Boxes

Boxes can be colors using `background`, `color`, or `type`:

<Playground>
	<Box align="center" background="primary">
		primary
	</Box>
	<Box align="center" type="danger">
		danger (type)
	</Box>
	<Box align="center" color="danger">
		danger (color)
	</Box>
</Playground>

### Boxes as rows

<Playground>
	<Box
		direction="row"
		padding="medium"
		style={{ border: '10px solid blue', width: '100px' }}
	>
		<Box margin={{ top: "small" }} padding="small" style={{ background: 'lightgrey' }} />
		<Box padding="medium" style={{ background: 'gray' }} />
	</Box>
</Playground>

### Various Padding and Margins

<Playground>
	<Box padding="large" style={{ background: 'blue' }}>
		<div style={{ backgroundColor: 'white' }}>padding</div>
	</Box>

	<div style={{ background: 'red' }}>
		<Box margin="large" padding="large" style={{ background: 'blue' }}>
			<div style={{ backgroundColor: 'white' }}>padding + margin</div>
		</Box>
	</div>
</Playground>

### Pull the box around content

<Playground>
	<Box padding="large" pull="left" style={{ background: 'blue' }}>
		<div style={{ backgroundColor: 'white' }}>Pulled</div>
	</Box>

	<Typography>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.</Typography>
</Playground>

### Create a simple grid

<Playground>
	<Box
		direction="row"
		padding="medium"
		gap="small"
		style={{ border: '10px solid blue' }}
	>
		<Box flex padding="medium" style={{ background: 'lightgrey' }} />
		<Box flex padding="medium" style={{ background: 'lightgrey' }} />
		<Box flex={3} padding="medium" style={{ background: 'lightgrey' }} />
	</Box>
</Playground>

<Playground>
	<Box
		direction="column"
		padding="medium"
		gap="small"
		style={{ border: '10px solid blue' }}
	>
		<Box flex padding="medium" style={{ background: 'lightgrey' }} />
		<Box flex padding="medium" style={{ background: 'lightgrey' }} />
		<Box flex={3} padding="medium" style={{ background: 'lightgrey' }} />
	</Box>
</Playground>

## Prop Types

<PropsTable of={BoxProps} />