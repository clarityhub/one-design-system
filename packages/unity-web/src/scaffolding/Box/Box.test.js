import React, { Fragment } from 'react';
import renderer from 'react-test-renderer';

import Box from './Box';

describe('<Box />', () => {
	it('renders text', () => {
		const node = (
			<Box>
				Children
			</Box>
		);
		const tree = renderer.create(node).toJSON();

		expect(tree).toMatchSnapshot();
	});
    
	it('renders types and colors', () => {
		const node = (
			<Fragment>
				<Box align="center" background="primary">
                    primary
				</Box>
				<Box align="center" type="danger">
                    danger (type)
				</Box>
				<Box align="center" color="danger">
                    danger (color)
				</Box>

			</Fragment>
		);

		const tree = renderer.create(node).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('renders paddings and margins', () => {
		const node = (
			<Fragment>
				<Box padding="large">
					<div>padding</div>
				</Box>

				<div>
					<Box margin="large" padding="large">
						<div>padding + margin</div>
					</Box>
				</div>
			</Fragment>
		);

		const tree = renderer.create(node).toJSON();

		expect(tree).toMatchSnapshot();
	});
    
	it('renders a simple grid', () => {
		const node = (
			<Box
				direction="row"
				padding="medium"
				gap="small"
			>
				<Box flex padding="medium" />
				<Box flex padding="medium" />
				<Box flex={3} padding="medium" />
			</Box>
		);

		const tree = renderer.create(node).toJSON();

		expect(tree).toMatchSnapshot();
	});
});