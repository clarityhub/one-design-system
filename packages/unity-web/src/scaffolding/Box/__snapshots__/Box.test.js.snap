// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Box /> renders a simple grid 1`] = `
.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding: 2rem 2rem 2rem 2rem;
}

.emotion-6 > *:not(:first-of-type) {
  margin-left: 1rem;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 2rem 2rem 2rem 2rem;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 3;
  -ms-flex: 3;
  flex: 3;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 2rem 2rem 2rem 2rem;
}

<div
  className="emotion-6 emotion-1"
>
  <div
    className="emotion-0 emotion-1"
  />
  <div
    className="emotion-0 emotion-1"
  />
  <div
    className="emotion-4 emotion-1"
  />
</div>
`;

exports[`<Box /> renders paddings and margins 1`] = `
Array [
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 4rem 4rem 4rem 4rem;
}

<div
    className="emotion-0 emotion-1"
  >
    <div>
      padding
    </div>
  </div>,
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 4rem 4rem 4rem 4rem;
  margin: 4rem 4rem 4rem 4rem;
}

<div>
    <div
      className="emotion-0 emotion-1"
    >
      <div>
        padding + margin
      </div>
    </div>
  </div>,
]
`;

exports[`<Box /> renders text 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="emotion-0 emotion-1"
>
  Children
</div>
`;

exports[`<Box /> renders types and colors 1`] = `
Array [
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background: #566ad2;
}

<div
    className="emotion-0 emotion-1"
  >
    primary
  </div>,
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background: #f5365c;
  color: #FFF;
}

<div
    className="emotion-0 emotion-1"
  >
    danger (type)
  </div>,
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  color: #f5365c;
}

<div
    className="emotion-0 emotion-1"
  >
    danger (color)
  </div>,
]
`;
